### PR TITLE
LCB: Rename standard library types.

### PR DESCRIPTION
**Pull request by peter-b**
The current standard library uses lower-case type names.  This has
some unfortunate interactions with the current syntax implementation;
for example, the use of an "index" type means that it's not possible
to add any syntax that uses "index" as a keyword.

Initial draft of spec for new naming convention for standard library
types.

[Rendered version](https://github.com/peter-b/livecode/blob/feature-lcbtyperename/docs/specs/lcb-type-naming.md)

### CODE REVIEW

**Comment by peter-b** [comment]
Please bikeshed this as quickly as possible, because I need to start implementing it this afternoon if possible.

**Comment by runrevmark** [comment]
Purple or Orange? ;)

**Comment by runrevmark** [comment]
UIntPtr / IntPtr? (Potentially consider Pointer -> Ptr ?)

Also, we should potentially have universal names for the binary-fp types bearing in mind we'll also have decimal arithmetic at some point - BinFlt32 / BinFlt64? (Which would allow DecFlt32 / DecFlt64 etc.)

**Comment by peter-b** [comment]
@runrevmark For now, I was aiming to simply provide names for stuff that we already have (in fact, I think the only exception to that rule is the fixed-width integer types that I'm not planning on actually implementing in this branch).

Can we consider additional native types at a later date / as required?

**Comment by runrevmark** [comment]
I think it would be useful to put other potential ones in the spec - for discussion and so we can see the scope of the naming policy (the more examples we have the easier it is to see whether we have the 'right' - for some definition of 'correct' - idea or not). (Just mark the ones which aren't implemented as 'TODO').

The reason I brought up the BinFlt32 suggestion was that we already have double/float implemented so should probably consider universal aliases. After all, LCIndex / CInt / LCInteger will all be aliases for the current 'int' foreign type.

**Comment by peter-b** [comment]
On the subject of `Pointer` -> `Ptr` rename: it seems to me that the aim of LiveCode is descriptiveness rather than conciseness; I think if I wanted to write very dense native code I'd write in C or Rust and then call it using Builder's FFI.

**Comment by peter-b** [comment]
Currently amending the specification to add other types as suggested.

**Comment by runrevmark** [comment]
@peter-b: I agree with descriptiveness - I'm sure there's a rule here somewhere...

IntPtr seems better somehow than IntPointer

BinFlt32 seems better somehow than BinaryFloat32

Perhaps Pointer can just be an exception :)

**Comment by runrevmark** [comment]
Btw, you never said whether you preferred purple or orange ;)

**Comment by peter-b** [comment]
@runrevmark: Orange all the way :bicyclist: :bicyclist: :bicyclist: 

Maybe we add the "verbose" versions to start with and then we can add abbreviated synonyms if we decide they're too unwieldy?

I've pushed a patch which adds some (verbose!) additional types to the spec.

**Comment by runrevmark** [comment]
Further to discussion offline - I've been disuaded from BinFloat... @peter-b made the point that hardware tends towards binary representations and Int32 implicitly implicitly implies a binary-encoded integer... Thus Float32 should also implicitly imply binary.

I also retract DecFloat in favour of Float32Dec - the latter is more consistent with the ZString naming.

**Comment by runrevmark** [comment]
I'm happy with this to be applied to the existing types we define :)

**Comment by peter-b** [comment]
@runrevmark: This branch now renames all existing uses of the old native types that I was able to find to use the new names, while keeping the deprecated old names around.

The main question now is to which module(s) to add the mixed-case versions of the type names (i.e. `string` → `String`).

**Comment by runrevmark** [comment]
Having seen this I wonder if we should (slightly breaking convention) make LCInteger, LCInt... The rationale is the potential confusion between 'high-level' Integer (LCIntegerRef) type and a POD 'Int' type. Also we are proposing Int32 and CInt - and probably Int (for platform-sized integer?).

In regards to the aliases - they should go in the module for the type... So string -> String in com.livecode.string.

**Comment by peter-b** [comment]
@runrevmark I've modified the compiler to use titlecase core types (e.g. `List`); lowercase core types (e.g. `list`) are still supported, but emit a deprecation warning.  I've update all of the standard library and test suite accordingly.  Because the existing widgets should still all work I haven't touched the `extensions/` hierarchy at all.  Finally, I've renamed `LCInteger` to `LCInt` and `LCUInteger` to `LCUInt`.

File: docs/specs/lcb-type-naming.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+* `CDouble`: C's `double` type.
+
+## LiveCode library types
+
+For foreign function calls into the LiveCode engine's C/C++ libraries,
+some special foreign types are required.
+
+* `LCIndex`, `LCUIndex`: Indices (e.g. into lists or strings)
+
+* `LCInteger`, `LCUInteger`: Integers
```
Comment (runrevmark) [inline docs/specs/lcb-type-naming.md]: This should probably be LCInt / LCUInt now :)

**Comment by runrevmark** [comment]
Reviewed-by: @runrevmark.

**Pull request merged**